### ISSUE TITLE
Track all customizations that makes llvm-spirv in intel/llvm

### ISSUE THREAD

**Issue by MrSidims** [intel/llvm#7592]
The goal is to keep intel/llvm's llvm-spirv on pair with https://github.com/KhronosGroup/SPIRV-LLVM-Translator as much as possible. I'm not expecting it to be always on pair, since some patches in https://github.com/KhronosGroup/SPIRV-LLVM-Translator requires appropriate update on devices' drivers in CI here, so the list should be updated.

Following patches are missing in this repository in comparison to https://github.com/KhronosGroup/SPIRV-LLVM-Translator :
1. [Add SPIR-V 1.4 checks](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/c5b3c8e3283b ). It's going to be resolved by https://github.com/intel/llvm/pull/7493 , but it's blocked by GPU driver updates on windows.
2. [Add an entry point wrapper around functions (llvm pass)](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/85815e725ce5bdc970b812b4bbff73d4b2a44046) . It depends on a patch 1.
3. [Translate function attributes (incl. parameter attrs) for entry point kernels](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/aded5afb04f02b1b057db96cd9a8a5e9d1ff47fc) . It depends on a patch 2.
4. [Fix the collection of entry point interfaces(https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1334) It depends on a patch 2.

This repository adds extra code:
1. [In SPIRVRegularizeLLVM.cpp](https://github.com/intel/llvm/blob/sycl/llvm-spirv/lib/SPIRV/SPIRVRegularizeLLVM.cpp#L415) we have a 'handling' of freeze instruction. We should either upstream W/A or implement a proper handling of the instruction in Khronos. See https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1140 for reference

This repository removes some code:
1. `FunctionParameterAttributeNoReadWrite` is removed from `SPIRVFuncParamAttrKind` map until we have devices drivers updated.

**Comment by MrSidims** [comment]
@vmaksimo please take a look and add notes if I'm missing something.

**Comment by bader** [comment]
> The goal is to keep intel/llvm's llvm-spirv on pair with https://github.com/KhronosGroup/SPIRV-LLVM-Translator as much as possible. I'm not expecting it to be always on pair, since some patches in https://github.com/KhronosGroup/SPIRV-LLVM-Translator requires appropriate update on devices' drivers in CI here, so the list should be updated.

In my opinion we should remove llvm-spirv sources from intel/llvm repository. It difficult to track the difference between sources in Khronos and intel repositories. Custom layout makes it hard to sync and maintain. I suggest we consider:
1. Add an intel branch in Khronos repository which should aim to be "always on par" with the main development branch but can be used to keep patches for Intel device. Non-SYCL compilers targeting Intel devices should benefit from this branch as well. If Khronos is against vendor branches in their repository, we can consider using a fork under intel organization.
2. Keep set of patches on top of Khronos main branch in intel/llvm repository instead of the copy of sources. Keeping track of changes should be a no-brainer.

**Comment by vmaksimo** [comment]
> please take a look and add notes if I'm missing something.

In this repo we also have `DecorationFuncParamDescINTEL` and `DecorationFuncParamKindINTEL`, which are not present in Khronos. Other things mentioned are the same as in #3686 issue, I'll close it as a duplicate.

**Comment by asudarsa** [comment]
> The goal is to keep intel/llvm's llvm-spirv on par with https://github.com/KhronosGroup/SPIRV-LLVM-Translator as much as possible. I'm not expecting it to be always on par, since some patches in https://github.com/KhronosGroup/SPIRV-LLVM-Translator requires appropriate update on devices' drivers in CI here, so the list should be updated.
> 
> **Following patches are missing** in this repository in comparison to https://github.com/KhronosGroup/SPIRV-LLVM-Translator :
> 
> * [ ]   1. [Add SPIR-V 1.4 checks](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/c5b3c8e3283b). It's going to be resolved by [[SPIR-V] Cherry-pick of "Add SPIR-V 1.4 checks" #7493](https://github.com/intel/llvm/pull/7493) , but it's blocked by GPU driver updates on windows.
> * [ ]   2. [Add an entry point wrapper around functions (llvm pass)](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/85815e725ce5bdc970b812b4bbff73d4b2a44046) . It depends on a patch 1.
> * [ ]   3. [Translate function attributes (incl. parameter attrs) for entry point kernels](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/aded5afb04f02b1b057db96cd9a8a5e9d1ff47fc) . It depends on a patch 2.
> * [ ]   4. [Fix the collection of entry point interfaces](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1334) It depends on a patch 2.
> 
> **This repository adds extra code:**
> 
> * [ ]   1. [In SPIRVRegularizeLLVM.cpp](https://github.com/intel/llvm/blob/sycl/llvm-spirv/lib/SPIRV/SPIRVRegularizeLLVM.cpp#L415) we have a 'handling' of freeze instruction. We should either upstream W/A or implement a proper handling of the instruction in Khronos. See [Unsupported instruction: freeze KhronosGroup/SPIRV-LLVM-Translator#1140](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/1140) for reference.
> * [ ]  2. [This PR](https://github.com/intel/llvm/pull/7632) adds a W/A to replace GEP on ID-related builtin GVs with `extractelement` instruction. Need to remove it, when we have a fix in SPIR-V consumer propagated to intel/llvm's CI as part of driver updates.
> 
> **This repository removes some code:**
> 
> * [ ]  1. [This PR](https://github.com/intel/llvm/pull/7669) adds W/A that removes `FunctionParameterAttributeNoReadWrite`, need to return it.

Hi @MrSidims 

I see PR in #1 has been merged. Can we proceed with the remaining pointers here?

Thanks

**Comment by bader** [comment]
@MrSidims, @asudarsa, https://github.com/intel/llvm/pull/11186 - it seems like we missed a commit made 3 years ago. Can you check if anything else is missing, please?

**Comment by AlexeySachkov** [comment]
Linking pulldown PR #11920 in here and in particular [this comment](https://github.com/intel/llvm/pull/11920#issuecomment-1820951384) and [this commit](https://github.com/intel/llvm/pull/11920/commits/3f8d7ed61ce0b2261f24d208a660580b32bb442c) - it should be reverted once KhronosGroup/SPIRV-LLVM-Translator#2224 is resolved. @jsji, @MrSidims: FYI

**Comment by jsji** [comment]
https://github.com/KhronosGroup/SPIRV-LLVM-Translator/commit/64fe75e388dc3b7 is reverted in https://github.com/intel/llvm/pull/12868 due to failures .
Failed Tests (3):
  SYCL :: NonUniformGroups/ballot_group_algorithms.cpp
  SYCL :: NonUniformGroups/opportunistic_group_algorithms.cpp
  SYCL :: NonUniformGroups/tangle_group_algorithms.cpp

"[7:34 PM] Sidorov, Dmitry
I was checking if we indeed pass runtime value and it's not some sort of missing constant propagation. Unfortunately id passed to OpGroupNonUniformBroadcast is always runtime value for SYCL headers. Hence we need to revert the patch. "

**Comment by asudarsa** [comment]
One more:
https://github.com/intel/llvm/commit/bbf3800adf1209d062ddb49555bb4f36120575ad

**Comment by jsji** [comment]
> One more: [bbf3800](https://github.com/intel/llvm/commit/bbf3800adf1209d062ddb49555bb4f36120575ad)

Yes, this is the one mentioned above https://github.com/intel/llvm/issues/7592#issuecomment-1972236053

**Comment by asudarsa** [comment]
Ah..Thanks. I missed that. Will be addressing these soon.

**Comment by maarquitos14** [comment]
https://github.com/intel/llvm/pull/15298 reverts the W/A that removes `FunctionParameterAttributeNoReadWrite`. Marked as solved in the description the corresponding item.

**Comment by uditagarwal97** [comment]
Added `disable-spirv-tools` LIT parameter to llvm-spirv tests to explicitly disable using spirv-tools in intel/llvm CI (https://github.com/intel/llvm/pull/16743). These changes will be reverted once the test failures are fixed.

**Comment by sarnex** [comment]
Lifetime intrinsic related customizations removed in https://github.com/intel/llvm/pull/22253
